pkg/plugin/trans.go

```diff
@@ -80,6 +84,40 @@ func (s verdictSet) MarshalLogObject(enc zapcore.ObjectEncoder) error {
 	return nil
 }
 
+// handleLastPermit updates reserved values of r.pod and r.node with the most
+// recent info about the last permit received by the agent. This allows a new
+// scheduler to learn about previous scheduler's state.
+//
+// Why the new scheduler needs this info? It's always possible for the scheduler
+// to get killed and immediately restart, without the agent believing there was
+// any disconnect, which could lead to unintentional over-committing of resources
+// from the Buffer values if too many agents request upscaling on the first
+// request to the scheduler.
+func (r resourceTransition[T]) handleLastPermit(lastPermit T) (verdict string) {
+	if lastPermit <= r.pod.Reserved {
+		r.node.Reserved -= r.pod.Reserved - lastPermit
+		r.pod.Reserved = lastPermit
+		verdict = fmt.Sprintf(
+			"pod reserved (%v -> %v), node reserved (%v -> %v)",
+			r.oldPod.reserved, r.pod.Reserved, r.oldNode.reserved, r.node.Reserved,
+		)
+	} else {
+		// This is an unexpected case that possible to happen in some unlikely scenarios such as:
+		//   1. Agent receives a permit from scheduler (let’s say it’s equal to `a` for a specific resource)
+		//   2. scheduler dies
+		//   3. vm bounds decrease (the max value is `b` and we have `b < a`).
+		//   4. new scheduler reads the cluster state and sets up the buffer values
+		//   => agent’s last permit is greater (`a`) than plugin’s reserved value (`b`)
+		// This might also happen in case of processing a stale request from an agent
+
+		verdict = fmt.Sprintf(
+			"unexpected last permit, no changes: last permit (%v) is greater than pod reserved (%v)",
+			lastPermit, r.pod.Reserved,
+		)
```

Comment:
I'd be worried that doing nothing here might leave us in a permanently degraded state. Thoughts?

Or, is this mostly "revert to previous behavior" because we call `handleRequested` afterwards anyways?

Reply:
I had the latter in mind, though we might later call this without having `handleRequested` be called afterwards. I don't think updating the reserved value in this case would be right, cause it's not the most recent value anymore.

Anyway, now I think it's better to enforce that `handleRequested` will be called later. I'm going to remove this function and make it part of `handleRequested`.